### PR TITLE
feat: Implement DataProxy

### PR DESCRIPTION
**Pull request by Ldoppea**
This PR implement the foundation for the DataProxy

The DataProxy is a new cozy-app meant to be embedded by other cozy-apps that need to access locally persisted data and heavy processes (like search indexing, maybe local AI in the future etc)

The goal is to mutualize the persisted data that will be accessed by multiple cozy-app that have all their own domain (as we use flat domain with slug for every cozy-app) and so cannot mutualize their local storage without additional mechanisms (this is why we want to implement the DataProxy)

Note that this cozy-app is not meant to be opened directly. Some work will be done to hide it from cozy-home and cozy-bar apps lists

Other cozy-app will be able to access the DataProxy through an iframe based API. This means that every cozy-app will be able to create an iframe pointing to the new DataProxy cozy-app

The impact is that multiple tabs may exist simultaneously and render the DataProxy cozy-app. So we want to ensure all the heavy work (i.e. search indexing) is done one single time. To make this possible, the DataProxy cozy-app will rely on a SharedWorker that will ensure only one process is instantiated for executing the code independently of how many tabs are open

With this SharedWorker mechanism, the following scenario will happen:
- If you open a cozy-home and a cozy-drive from the same Cozy instance in two different tabs
  - Those tabs will use the same SharedWorker process
  - This process will stay up until ALL the tabs are closed
- If you update the DataProxy
  - All the new tabs on cozy-home or cozy-drive will use a new SharedWorker process using the new code
  - All the existing tabs (opened before the update) will still use the old SharedWorker process using the old code
- Each different cozy instance will use a different process

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import { useSharedWorker } from '@/dataproxy/worker/useSharedWorker'
+
+import './App.css'
+
+const log = Minilog('🖼️ [DataProxy main]')
+Minilog.enable()
+
+const App = (): JSX.Element => {
+  const client = useClient()
+  const { worker, workerState } = useSharedWorker()
```
Comment (paultranvan) [inline src/App.tsx:17]: `worker` looks useless here?

File: src/doctypes/index.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  homeworks: {
+    doctype: 'io.cozy.calendar.todos'
+  },
+  grades: {
+    doctype: 'io.cozy.timeseries.grades'
+  },
+  presence: {
+    doctype: 'io.cozy.calendar.presence'
+  }
+}
```
Comment (paultranvan) [inline src/doctypes/index.ts:15]: To remove

File: src/dataproxy/worker/SharedWorkerProvider.tsx [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+      }
+    )
+    const [tabCount, setTabCount] = useState(0)
+    const client = useClient()
+
+    useEffect(() => {
+      if (!client) return
+
+      const doAsync = async (): Promise<void> => {
+        log.log('Init SharedWorker')
```
Comment (paultranvan) [inline src/dataproxy/worker/SharedWorkerProvider.tsx:47]: What is a `log.log` compared to `log.debug` ? :thinking:

**Review by paultranvan — APPROVED** [review]
Really great work :clap: for setting up this

File: src/dataproxy/worker/SharedWorkerProvider.tsx (hunk 3 above)
Comment (Ldoppea) [reply src/dataproxy/worker/SharedWorkerProvider.tsx:47]: No difference, they are mapped to the same method. I just replace `console` with `log`.

I'll replace this

File: src/App.tsx (hunk 1 above)
Comment (Ldoppea) [reply src/App.tsx]: [Fixed](https://github.com/cozy/cozy-web-data-proxy/compare/706f11d012a29e5971ca7e3b9ca740d220b21e05..d4c89625494b97d98a279dca53386d11c881e736)

File: src/dataproxy/worker/SharedWorkerProvider.tsx (hunk 3 above)
Comment (Ldoppea) [reply src/dataproxy/worker/SharedWorkerProvider.tsx]: [Fixed](https://github.com/cozy/cozy-web-data-proxy/compare/706f11d012a29e5971ca7e3b9ca740d220b21e05..d4c89625494b97d98a279dca53386d11c881e736)

**Pull request merged**